_posts/2019-02-06-Age-Targeting-Part2.md

```diff
@@ -0,0 +1,255 @@
+The weights are calculated through the following code.
+It produces a weight column in the dataframe as output.
+
+```python
+def get_weight_function(label_counts):
+    max_cnt = float(max(label_counts.values()))
+    label_weight = {k:max_cnt/v for k, v in label_counts.items()}
+
+    def get_value(x):
+        return label_weight[x]
+    return get_value, label_weight
+
+row_counts = df.groupby(label_col).count().collect()
+label_counts = {row[label_col]:row["count"] for row in row_counts}           
+get_weight, label_weight = get_weight_function(label_counts)
+df = df.withColumn(weight_col, F.udf(get_weight, DoubleType())(label_col))
+```
+
+
+**Feature creation**<br/>
+The question to keep in mind while creating the feature is: how do people of different age use their smart devices?
+The informations which contain a difference in the usage of the mobile for different classes are the ideal features to learn about the age of the user.
+Only the ones which contribute to improve the prediction are kept in the final model.
+
+* **Time distribution**<br/>
+We calculate the hourly and weekly usage distribution, namely the fraction of the total time spent on apps split by hour or by day of the week.
+On the plot we can see the average of the hourly fraction calculated over every devices per each classes.
+The trends follow the expectations: the younger tend to spend more time on the apps late in the evening, while the older like to be active earlier in the morning.
+<p align="center"> <img src="../images/Age-Targeting/parallel_hour.png"> </p>
+
+* **The apps**<br/>
+We collect the apps used over the last month for each user.
+This feature is directly related to the different kind of usages for a smart devices.
+As we will see later in the post, it has the strongest relation to the age.
+To be passed as a feature in the machine learning model, we transform the list of used apps to a [bag of words](https://en.wikipedia.org/wiki/Bag-of-words_model) via the CountVectorizer class in spark.
+In this way every apps constitute a single feature.
+
+```python
+from pyspark.ml.feature import CountVectorizer, Normalizer
+
+apps_count_vectorizer = CountVectorizer(inputCol=feature_col, 
+                                        outputCol=count_vectorizer_col, 
+                                        minDF=min_df)
+apps_normalizer = Normalizer(p=2.0, 
+                             inputCol=count_vectorizer_col, 
+                             outputCol=normalizer_col)
+```
+The minDF parameter of the CountVectorizer class is used to cut off apps visited by few devices to decrease the noise in the dataset and the chance to overfit.
+
+* **Make and model**<br/>
+Both make and model can be expressed as categorical features and treated via [One Hot Encoding](https://en.wikipedia.org/wiki/One-hot).
+Due to the high cardinality of the models we map the low frequent occurrences into a single index to reduce the sparsity of the data and reduce the overfitting.
```

Comment:
OK, now you are getting serious. I like the sentence, but only Engineers will be able to understand it :)